pkg/backends/calico/routes.go

```diff
@@ -15,30 +15,240 @@ package calico
+		// attempt 2: in cluster config
+		if cfg, err = rest.InClusterConfig(); err != nil {
+			return nil, err
+		}
+	}
+	client, err := kubernetes.NewForConfig(cfg)
+	if err != nil {
+		return nil, err
+	}
+
+	// set up services informers
+	svcWatcher := cache.NewListWatchFromClient(client.Core().RESTClient(), "services", "", fields.Everything())
+	svcHandler := cache.ResourceEventHandlerFuncs{AddFunc: rg.onSvcAdd, UpdateFunc: rg.onSvcUpdate, DeleteFunc: rg.onSvcDelete}
+	rg.svcIndexer, rg.svcInformer = cache.NewIndexerInformer(svcWatcher, &v1.Service{}, 0, svcHandler, cache.Indexers{})
+
+	// set up endpoints informers
+	epWatcher := cache.NewListWatchFromClient(client.Core().RESTClient(), "endpoints", "", fields.Everything())
+	epHandler := cache.ResourceEventHandlerFuncs{AddFunc: rg.onEPAdd, UpdateFunc: rg.onEPUpdate, DeleteFunc: rg.onEPDelete}
+	rg.epIndexer, rg.epInformer = cache.NewIndexerInformer(epWatcher, &v1.Endpoints{}, 0, epHandler, cache.Indexers{})
+
+	log.Debug("-NewRouteGenerator")
 	return
 }
 
 func (rg *routeGenerator) Start() (err error) {
 	// MVP implementation: read CIDRs to advertise,
 	// comma-separated, from an environment variable
 	// CALICO_STATIC_ROUTES.
-	routeString := os.Getenv("CALICO_STATIC_ROUTES")
-	cidrs := []string{}
-	for _, route := range strings.Split(routeString, ",") {
-		cidr := strings.TrimSpace(route)
-		if cidr != "" {
-			cidrs = append(cidrs, cidr)
+	ch := make(chan struct{})
+	go rg.svcInformer.Run(ch)
+	go rg.epInformer.Run(ch)
+
+	/*
+		routeString := os.Getenv("CALICO_STATIC_ROUTES")
+		cidrs := []string{}
+		for _, route := range strings.Split(routeString, ",") {
+			cidr := strings.TrimSpace(route)
+			if cidr != "" {
+				cidrs = append(cidrs, cidr)
+			}
+		}
+		rg.client.updateRoutes(cidrs)
+	*/
+	return
+}
+
+// setRouteForSvc handles the main logic to check if a specified service or endpoint
+// should have its route advertised by the node running this code
+func (rg *routeGenerator) setRouteForSvc(svc *v1.Service, ep *v1.Endpoints) {
+	log.Debug("+setRouteForSvc")
+	rg.Lock()
+	defer rg.Unlock()
+
+	// ensure both are not nil
+	if svc == nil && ep == nil {
+		log.Warn("setRouteForSvc: both service and endpoint cannot be nil, passing...")
+		return
+	}
+
+	var (
+		key string
+		err error
+	)
+
+	if svc == nil {
+		// ep received but svc nil
+		// get key
+		key, err = cache.MetaNamespaceKeyFunc(ep)
+		if err != nil {
+			log.WithField("svc", svc.Name).WithError(err).Warn("setRouteForSvc: error on retrieving key for endpoint, passing")
+			return
+		}
+		// get svc
+		svcIface, exists, err := rg.svcIndexer.GetByKey(key)
+		if err != nil {
+			log.WithField("key", key).WithError(err).Warn("setRouteForSvc: error on retrieving service for key, passing")
+			return
+		} else if !exists {
+			log.WithField("key", key).Warn("setRouteForSvc: service for key not found, passing")
+			return
+		}
+		svc = svcIface.(*v1.Service)
+	} else if ep == nil {
+		// svc received but ep nil
```

Comment:
Another place we could perhaps do with breaking out into sub-functions.

e.g.

```go
func (rg RouteGenerator) getServiceForEndpoints() *v1.Service {}
func (rg RouteGenerator) getEndpointsForService() *v1.Endpoints {}
```

This would also mean we could tighten up the contract for this function. For example, we could say that the calling code _needs_ to pass both a Service and an Endpoints object, and then this function wouldn't need to do any storage access, making it easier to unit test.